### ISSUE TITLE
The differences between the pure inference results of the Torch model and the ONNX model are quite significant.

### ISSUE THREAD

**Issue by yjmm10** [Peterande/D-FINE#208]
Utilizing the official ONNX conversion script, the preprocessing step was removed from the complete inference process. Since the post-processing is encapsulated on the model side, temporarily we do not consider extracting it. Instead, we use random numbers to conduct inference on the PyTorch model and the ONNX model. The results are quite different. Below are the comparison scripts and results I have conducted.

```python
"""
Copyright (c) 2024 The D-FINE Authors. All Rights Reserved.
"""

import torch
import torch.nn as nn
import sys
import os
import onnxruntime as ort
import numpy as np
import argparse

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../')))
from src.core import YAMLConfig

class Model(nn.Module):
    def __init__(self, cfg):
        super().__init__()
        self.model = cfg.model.deploy()
        self.postprocessor = cfg.postprocessor.deploy()

    def forward(self, images,orig_target_sizes):
        outputs = self.model(images)
        outputs = self.postprocessor(outputs, orig_target_sizes)
        return outputs

class TorchModel:
    """PyTorch模型加载和推理类"""
    
    def __init__(self, config_path, checkpoint_path, device='cpu'):
        """初始化模型类
        
        Args:
            config_path: 配置文件路径
            checkpoint_path: 模型权重路径
            device: 运行设备，默认为CPU
        """
        # 加载配置文件
        self.cfg = YAMLConfig(config_path, resume=checkpoint_path)
        
        # 关闭预训练
        if 'HGNetv2' in self.cfg.yaml_cfg:
            self.cfg.yaml_cfg['HGNetv2']['pretrained'] = False
            
        # 加载权重
        checkpoint = torch.load(checkpoint_path, map_location='cpu')
        if 'ema' in checkpoint:
            state = checkpoint['ema']['module']
        else:
            state = checkpoint['model']
            
        # 加载训练模式状态并转换为部署模式
        self.cfg.model.load_state_dict(state)
        
        # 创建模型实例
        self.model = Model(self.cfg).to(device)
        self.model.eval()  # 设置为评估模式
        
    def __call__(self, images, orig_target_sizes=None):
        """模型推理函数，仅执行模型前向传播，不包含前后处理
        
        Args:
            images: 输入图像张量
            
        Returns:
            模型原始输出结果
        """
        if orig_target_sizes is None:
            orig_target_sizes = torch.tensor([[640, 640]])
        with torch.no_grad():
            outputs = self.model(images,orig_target_sizes)
        return outputs


# onnx模型加载，
class ONNXModel:
    """ONNX模型加载和推理类"""
    
    def __init__(self, model_path):
        """初始化ONNX模型类
        
        Args:
            model_path: ONNX模型文件路径
        """
        import onnxruntime as ort
        # 加载ONNX模型
        self.sess = ort.InferenceSession(model_path)
        
    def __call__(self, images, orig_target_sizes=None):
        """模型推理函数
        
        Args:
            images: 输入图像张量
            orig_target_sizes: 原始目标尺寸，默认为640x640
            
        Returns:
            模型输出结果
        """
        if orig_target_sizes is None:
            orig_target_sizes = torch.tensor([[640, 640]])
            
        # 执行ONNX模型推理
        output = self.sess.run(
            output_names=None,
            input_feed={"images": images.numpy(), "orig_target_sizes": orig_target_sizes.numpy()},
        )
        
        return output

def compare_models(config_path, torch_checkpoint_path, onnx_model_path, device='cpu'):
    """
    初始化PyTorch和ONNX模型，并比较它们在相同输入下的推理结果
    
    Args:
        config_path: 配置文件路径
        torch_checkpoint_path: PyTorch模型权重路径
        onnx_model_path: ONNX模型文件路径
        device: 运行设备，默认为CPU
    """
    # 初始化模型
    torch_model = TorchModel(config_path, torch_checkpoint_path, device)
    onnx_model = ONNXModel(onnx_model_path)
    
    # 创建随机输入
    batch_size = 1
    channels = 3
    height = 640
    width = 640
    
    # 设置随机种子以确保可重复性
    torch.manual_seed(42)
    random_input = torch.rand(batch_size, channels, height, width, dtype=torch.float32)
    orig_target_sizes = torch.tensor([[height, width]])
    
    # 运行PyTorch模型
    torch_output = torch_model(random_input, orig_target_sizes)
    
    # 运行ONNX模型
    onnx_output = onnx_model(random_input, orig_target_sizes)
    
    # 比较结果
    print("PyTorch模型输出类型:", type(torch_output))
    print("ONNX模型输出类型:", type(onnx_output))
    
    # 如果输出是字典或列表，打印其结构
    if isinstance(torch_output, dict):
        print("\nPyTorch输出键:", torch_output.keys())
    elif isinstance(torch_output, (list, tuple)):
        print("\nPyTorch输出长度:", len(torch_output))
        
    if isinstance(onnx_output, (list, tuple)):
        print("ONNX输出长度:", len(onnx_output))
    
    # 使用numpy.allclose进行精确比较
    try:
        # 根据错误信息，我们知道两个模型都返回了3个元素的输出
        # 分别处理每个输出元素进行比较
        if isinstance(torch_output, tuple) and isinstance(onnx_output, list) and len(torch_output) == 3 and len(onnx_output) == 3:
            print("\n分别比较三个输出元素:")
            
            # 比较第一个元素 (labels)
            torch_labels = torch_output[0]
            onnx_labels = torch.tensor(onnx_output[0])
            print(f"标签形状 - PyTorch: {torch_labels.shape}, ONNX: {onnx_labels.shape}")
            labels_match = np.allclose(torch_labels.cpu().numpy(), onnx_labels.cpu().numpy(), rtol=1e-5, atol=1e-5)
            print(f"标签数值是否匹配: {labels_match}")
            
            # 比较第二个元素 (boxes)
            torch_boxes = torch_output[1]
            onnx_boxes = torch.tensor(onnx_output[1])
            print(f"边界框形状 - PyTorch: {torch_boxes.shape}, ONNX: {onnx_boxes.shape}")
            boxes_match = np.allclose(torch_boxes.cpu().numpy(), onnx_boxes.cpu().numpy(), rtol=1e-5, atol=1e-5)
            print(f"边界框数值是否匹配: {boxes_match}")
            
            # 比较第三个元素 (scores)
            torch_scores = torch_output[2]
            onnx_scores = torch.tensor(onnx_output[2])
            print(f"置信度形状 - PyTorch: {torch_scores.shape}, ONNX: {onnx_scores.shape}")
            scores_match = np.allclose(torch_scores.cpu().numpy(), onnx_scores.cpu().numpy(), rtol=1e-5, atol=1e-5)
            print(f"置信度数值是否匹配: {scores_match}")
            
            # 打印总体匹配结果
            print(f"\n总体匹配结果: {labels_match and boxes_match and scores_match}")
            
            # 如果不匹配，打印差异的统计信息
            if not (labels_match and boxes_match and scores_match):
                print("\n差异统计信息:")
                if not labels_match:
                    labels_diff = np.abs(torch_labels.cpu().numpy() - onnx_labels.cpu().numpy())
                    print(f"标签最大差异: {np.max(labels_diff)}, 平均差异: {np.mean(labels_diff)}")
                
                if not boxes_match:
                    boxes_diff = np.abs(torch_boxes.cpu().numpy() - onnx_boxes.cpu().numpy())
                    print(f"边界框最大差异: {np.max(boxes_diff)}, 平均差异: {np.mean(boxes_diff)}")
                
                if not scores_match:
                    scores_diff = np.abs(torch_scores.cpu().numpy() - onnx_scores.cpu().numpy())
                    print(f"置信度最大差异: {np.max(scores_diff)}, 平均差异: {np.mean(scores_diff)}")
            
            # 打印每个元素的前几个值进行直观比较
            num_samples = 5  # 控制显示的样本数量
            
            print("\n标签值示例:")
            print(f"PyTorch: {torch_labels[0][:num_samples]}")
            print(f"ONNX: {onnx_labels[0][:num_samples]}")
            
            print("\n边界框值示例:")
            print(f"PyTorch: {torch_boxes[0][0]}")
            print(f"ONNX: {onnx_boxes[0][0]}")
            
            print("\n置信度值示例:")
            print(f"PyTorch: {torch_scores[0][:num_samples]}")
            print(f"ONNX: {onnx_scores[0][:num_samples]}")
        else:
            print("输出结构不符合预期，无法直接比较")
            print(f"PyTorch输出类型: {type(torch_output)}")
            print(f"ONNX输出类型: {type(onnx_output)}")
            
            if isinstance(torch_output, (list, tuple)):
                for i, item in enumerate(torch_output):
                    print(f"PyTorch输出[{i}]类型: {type(item)}")
                    
            if isinstance(onnx_output, (list, tuple)):
                for i, item in enumerate(onnx_output):
                    print(f"ONNX输出[{i}]类型: {type(item)}")
    except Exception as e:
        print(f"比较输出时出错: {e}")
        print("请手动检查两个模型的输出结构和值")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="比较PyTorch和ONNX模型的推理结果")
    parser.add_argument("--config", type=str, default="configs/*.yaml", help="配置文件路径")
    parser.add_argument("--torch_checkpoint", type=str, default="best_stg1.pth", help="PyTorch模型权重路径")
    parser.add_argument("--onnx_model", type=str, default="*.onnx", help="ONNX模型文件路径")
    parser.add_argument("--device", type=str, default="cpu", help="运行设备，默认为CPU")
    
    args = parser.parse_args()
    
    compare_models(
        args.config,
        args.torch_checkpoint,
        args.onnx_model,
        args.device
    )
```

![Image](https://github.com/user-attachments/assets/ca7f0173-27e0-496a-a1c1-1a1ea4f14c8b)

**Comment by Peterande** [comment]
Has the ONNX file you are using been processed with fp16? If it has been processed, it indicates that it may be due to environmental reasons. You can check it according to the relevant issue, such as https://github.com/Peterande/D-FINE/issues/44.

**Comment by S130111** [comment]
> 你使用的 ONNX 文件是否用 fp16 处理过？如果已经处理过，说明可能是环境原因，可以根据相关 issue 进行检查，比如[#44](https://github.com/Peterande/D-FINE/issues/44)。

定位到了 是由于 PIL resize 与cv resize 不同造成的差异，十分感谢

**Comment by hua1024** [comment]
I encountered the same issue. Have you found a solution?

**Comment by Nuzhny007** [comment]
@hua1024 do you export to onnx with 16 opset or 17 and higher? The default 16 is too low for TensorRT:
https://github.com/Peterande/D-FINE/blob/3c6711c8a1219eced79aba957875661356faa415/tools/deployment/export_onnx.py#L78

**Comment by hua1024** [comment]
> [@hua1024](https://github.com/hua1024) do you export to onnx with 16 opset or 17 and higher? The default 16 is too low for TensorRT:
> 
> [D-FINE/tools/deployment/export_onnx.py](https://github.com/Peterande/D-FINE/blob/3c6711c8a1219eced79aba957875661356faa415/tools/deployment/export_onnx.py#L78)
> 
> Line 78 in [3c6711c](/Peterande/D-FINE/commit/3c6711c8a1219eced79aba957875661356faa415)
> 
>  opset_version=16,

The final reason found was that there were differences between the preprocessing used during training and the preprocessing used during inference，torch use resize，onnx use resize+pad

**Comment by yjmm10** [comment]
> > [@hua1024](https://github.com/hua1024) do you export to onnx with 16 opset or 17 and higher? The default 16 is too low for TensorRT:
> > [D-FINE/tools/deployment/export_onnx.py](https://github.com/Peterande/D-FINE/blob/3c6711c8a1219eced79aba957875661356faa415/tools/deployment/export_onnx.py#L78)
> > Line 78 in [3c6711c](/Peterande/D-FINE/commit/3c6711c8a1219eced79aba957875661356faa415)
> > opset_version=16,
> 
> The final reason found was that there were differences between the preprocessing used during training and the preprocessing used during inference，torch use resize，onnx use resize+pad

Yes, the main issue is still the difference in post-processing methods. There are indeed some accuracy errors between pt and onnx, around 0.1-0.2%, but I think this is understandable. Thank you.
